### PR TITLE
bridge: Fix text channels with split UTF8 chars across frames

### LINKED ISSUE #19235: text channels drop frames with split UTF-8 multi-byte characters
Our tests occasionally fail with unexpected messages

    invalid non-UTF8 @data passed as text to web_socket_connection_send()

Today it happened [here](https://artifacts.dev.testing-farm.io/bef4337d-49f1-4670-8c28-1487b8f7aaad/) in `TestJournal.testAbrtReport`.  There is no other information about that, so we need to collect some instances to identify some patterns.

It may also help to actually show the data, so that we get a clue what happens.

 - [testAbrtReport rawhide on TF](https://artifacts.dev.testing-farm.io/bd093e0d-c09a-40ae-ad55-027006963065/)
 - [testAbrtReportCancel f38 on our CI](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19152-20230825-130641-f8cd4822-fedora-38-other/log.html#122)
 - [testAbrtSegv f38 on our CI](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19246-20230828-054020-5d786ab8-fedora-38-devel/log.html#288)
 - [testAbrtSegv f39 on our CI](https://cockpit-logs.us-east-1.linodeobjects.com/pull-5171-20230829-063656-e518ac35-fedora-39-other-cockpit-project-cockpit/log.html#123)
 - [testHiddenRaid on f39](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19254-20230829-091131-245b0fad-fedora-38-devel/log.html#188)

:arrow_up: By today (June 2024) the above logs are all gone, but https://github.com/cockpit-project/cockpit-podman/issues/1733 is a recent case with recent logs which is very easy to reproduce -- but it won't be any more in a few days. See below, @martinpitt  captured this bug in a unit test.

### PR DESCRIPTION
**Pull request by martinpitt**
Fixes #19235

### CODE REVIEW

File: src/cockpit/channel.py [hunk 1]
```diff
@@ -275,7 +277,20 @@ def send_data(self, data: bytes) -> bool:
...
+            # as these can't be sent to a WebSocket (and are often confusing for text streams as well)
+            if self.decoder is None:
+                self.decoder = codecs.getincrementaldecoder('utf-8')(errors='strict')
+            try:
+                block = self.decoder.decode(data)
+            except UnicodeDecodeError as exc:
+                raise ChannelError('protocol-error', message=str(exc)) from exc
+            self.send_channel_data(self.channel, block.encode())
+            # note: It is possible that after the last send_data() call there is a residual state in the decoder.
+            # That can only happen if a text channel receives invalid/truncated UTF-8 data. So we just don't send it.
```
Comment (martinpitt) [inline src/cockpit/channel.py:293]: Let's do something about this: If there's residual state in `self.close()` it could at least log a warning. However, this should only happen if the channel was closed without a `problem` code, as errors can happen at any time. @allisonkarlitskaya WDYT?

Comment (martinpitt) [reply src/cockpit/channel.py:293]: Thanks, I wasn't sure about that -- but right, this is the reading side -- we'd always get message-message-message-done.

**Review by allisonkarlitskaya — CHANGES_REQUESTED** [review]
I would have structured this a bit differently:

 - `.send_data(data: bytes)`: leave it alone, but add a comment that the caller is responsible for sending properly-formed utf8 data on non-binary channels.
 - `.send_text(data: str)`: essentially `self.send_bytes(data.encode())`
 - `.send_magic_stuff(data: bytes)`: if this is a binary channel, just `.send_data()`, if it's text then incrementally decode `data`, passing it to `.send_text()`.

I feel like there's a few places where we would like to use `.send_text()` — grepping around in the channels shows a few cases of `.send_data(x.encode())` and many cases of `.send_data(b'')` which are probably meant as text.

It might also make sense to rename the existing `.send_data()` to `.send_bytes()` and then the new "magic" function would be `.send_data()`.  I'm not too worried about the names.

The main thrust of this is that someone calling `self.send_data(x.encode())` on a text channel should not then have their data re-decoded, only to be encoded yet again... There should be a .send_text() to avoid that, and this is what the "magic" function should also call internally on the "string" that it gets from the incremental decoder.

File: src/cockpit/channel.py (hunk 1 above)
Comment (martinpitt) [reply src/cockpit/channel.py]: I did that now.

File: pkg/base1/test-http.js [hunk 2]
```diff
@@ -131,6 +131,36 @@ QUnit.test("streaming", assert => {
...
+            .get("/mock/truncated-utf8")
+            .stream(block => { received += block })
+            .then(() => assert.ok(false, "should not have succeeded"))
+            // does not include the first byte of é
+            .catch(ex => {
+                // does not include the first byte of é
+                assert.equal(received, "initialfirst half ", "received expected data");
+                assert.equal(ex.problem, "protocol-error", "error code");
+                // FIXME: the message is also "protocol-error", huh?
+                assert.equal(JSON.stringify(ex), "Received incomplete UTF-8 message before EOF", "error message");
```
Comment (martinpitt) [inline pkg/base1/test-http.js:159]: @allisonkarlitskaya do you have any idea what's wrong here? Of course the `stringify()` is complete nonsense, it's just for debugging. The channel sends it alright:
```
cockpit.protocol-DEBUG: sending control message {'message': 'Received incomplete UTF-8 message before EOF', 'problem': 'protocol-error'} {'command': 'close', 'channel': '1:2'}
```
but the assertion doesn't see it correctly:
```
g: not ok 4 - error message, expected: "Received incomplete UTF-8 message before EOF", got: "{\"problem\":\"protocol-error\",\"message\":\"protocol-error\"}", test: truncated UTF8 frame
```
WTH happens here?

**Comment by martinpitt** [comment]
Reworked as above to split into `send_{bytes,text,data}()`. I still have that pesky exception message problem from above, but that's too hairy for the rest of the day. Let's get an opinion from the bots in the meantime.

File: pkg/base1/test-http.js (hunk 2 above)
Comment (martinpitt) [reply pkg/base1/test-http.js:159]: Stupid bug in cockpit.js, found it and made a separate commit.

**Review by allisonkarlitskaya — CHANGES_REQUESTED** [review]
I did a survey of channels that are actually sending data other than .send_json():

 - fsread provides possibly-split-character data via its `do_yield_data()` generator.  It would be easy to change the `open()` call in there to change its mode based on if the channel is binary or not
 - packages is similar, but takes a different path: `AsyncChannel.sendfile()`.  It would probably be more difficult to adapt.
 - HTTP is difficult: I think there's no way to get text data out of the HTTP client API.
 - the echo channel sends whatever we send to it.  It's a good candidate for send_bytes().
 - stream goes via `ProtocolChannel`.  Your first patch moved that to `.send_bytes()` and didn't move it back to `.send_data()` later — that means a stream channel opened in text mode could still send invalid utf8.

So in summary: although we could modify fsread and packages to give us text data, because of HTTP and because of the stream channel, I think we're more or less forced to do the incremental coder for ourselves, at least once, and Channel seems to be the sensible spot for it.

File: pkg/lib/cockpit.js [hunk 3]
```diff
@@ -3896,7 +3896,7 @@ function factory() {
 
                 if (options.problem) {
                     http_debug("http problem: ", options.problem);
-                    dfd.reject(new BasicError(options.problem));
+                    dfd.reject(new BasicError(options.problem, options.message));
```
Comment (cockpituous) [inline pkg/lib/cockpit.js:3899]: This added line is not executed by any test.

File: src/cockpit/channel.py [hunk 4]
```diff
@@ -219,6 +221,9 @@ def ready(self, **kwargs: JsonValue) -> None:
         self.send_control(command='ready', **kwargs)
 
     def done(self) -> None:
+        # any residue from partial send_data() frames?
+        if self.decoder and self.decoder.getstate()[0]:
```
[the remark this answers is not in the document]
Comment (martinpitt) [reply src/cockpit/channel.py:225]: I'm aware of that, but that does more work than checking the state. It is a bit more consistent though. No strong opinion.

**Comment by martinpitt** [comment]
> the echo channel sends whatever we send to it. It's a good candidate for send_bytes().

I thought the same -- it's a case of GIGO, and it's unlikely that it's accidentally fed with invalid UTF-8. And if so, that'd ideally already fail on the sender side, not on the receiver. So I left it alone.

> that means a stream channel opened in text mode could still send invalid utf8.

Oops, will fix. Thanks!

File: pkg/lib/cockpit.js (hunk 3 above)
Comment (cockpituous) [inline pkg/lib/cockpit.js:3899]: This added line is not executed by any test.

File: src/cockpit/channel.py [hunk 5]
```diff
@@ -285,6 +289,17 @@ def send_data(self, data: bytes) -> bool:
 
         return self._out_sequence < self._out_window
 
+    def send_text(self, data: str) -> bool:
+        """Send UTF-8 string data and handle book-keeping for flow control.
+
+        Similar to `send_bytes`, but for text data.  The data is sent as UTF-8 encoded bytes.
+        """
+        return self.send_bytes(data.encode("UTF-8"))
```
Comment (allisonkarlitskaya) [inline src/cockpit/channel.py:326]: Sometimes you use `.decode()` and `.encode()` with an argument, sometimes without.  `utf-8` is the default value, so I think it always makes sense without.

**Review by allisonkarlitskaya — CHANGES_REQUESTED** [review]

File: src/cockpit/channels/metrics.py [hunk 6]
```diff
@@ -160,7 +160,7 @@ def send_updates(self, samples: Samples, last_samples: Samples):
             self.send_meta(samples, timestamp)
 
         self.last_timestamp = self.next_timestamp
-        self.send_data(json.dumps([data]).encode())
```
[the remark this answers is not in the document]
Comment (martinpitt) [reply src/cockpit/channels/metrics.py:163]: I first thought the same, but `send_json()` is meant to get the keys/values as kwargs. Of course we *could* do `send_json(**data)`, but converting it back and forth feels weird/brittle to me.

File: src/cockpit/channel.py [hunk 7]
```diff
@@ -125,6 +126,7 @@ def do_control(self, command: str, message: JsonObject) -> None:
             self._ack_bytes = get_enum(message, 'send-acks', ['bytes'], None) is not None
             self.group = get_str(message, 'group', 'default')
             self.is_binary = get_enum(message, 'binary', ['raw'], None) is not None
+            self.decoder: 'codecs.IncrementalDecoder | None' = None
```
[the remark this answers is not in the document]
Comment (martinpitt) [reply src/cockpit/channel.py:129]: How do you mean? Putting it into `__init__`? In case you meant a class attribute: that would be very much wrong of course.

[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply src/cockpit/channel.py:129]: Hanging default values directly on the class is very common practice and we do it all over the place.  Python even invented `ClassVar` that you need to explicitly use to annotate the cases of that which you actually intend to be treated as actual class attributes.

File: src/cockpit/channels/metrics.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (allisonkarlitskaya) [reply src/cockpit/channels/metrics.py:163]: Ah ya, let's skip this.  I knew there was a reason why this one wasn't the same...

File: src/cockpit/channel.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply src/cockpit/channel.py:129]: Sorry, I hate this behavior. I really want to keep it here, where it is crystal clear and explicit that it is an instance variable.

[the remark this answers is not in the document]
Comment (martinpitt) [reply src/cockpit/channel.py]: I moved the type declaration into the class, which does make sense.

FTR, I consider the initialization of `channel` and `group` to `''` at the class level a potential bug and unsafe practice. Accessing these fields in a non-open channel would better raise an `AttributeError` than silently succeeding and misbehaving. But that's the "agree to disagree" which we elaborated at length in the chat.

File: src/cockpit/channel.py [hunk 8]
```diff
@@ -219,6 +221,12 @@ def ready(self, **kwargs: JsonValue) -> None:
         self.send_control(command='ready', **kwargs)
 
     def done(self) -> None:
+        # any residue from partial send_data() frames?
+        if self.decoder:
+            try:
+                self.decoder.decode(b'', final=True)
+            except UnicodeDecodeError as exc:
```
[the remark this answers is not in the document]
Comment (martinpitt) [reply src/cockpit/channel.py]: I'm not a fan of exposing the `final` flag to the external API, as that's just an internal helper. I moved it into a separate private method.

File: src/cockpit/channel.py (hunk 5 above)
Comment (martinpitt) [reply src/cockpit/channel.py]: Done as a separate commit throughout the code.

File: pkg/lib/cockpit.js (hunk 3 above)
Comment (cockpituous) [inline pkg/lib/cockpit.js:3899]: This added line is not executed by any test.

File: src/cockpit/channel.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply src/cockpit/channel.py]: See below, reworked into a helper, so both places now use the same code.

File: pkg/lib/cockpit.js (hunk 3 above)
Comment (cockpituous) [inline pkg/lib/cockpit.js:3899]: This added line is not executed by any test.

**Comment by martinpitt** [comment]
@allisonkarlitskaya gentle review ping?

**Review by allisonkarlitskaya — APPROVED** [review]
Thanks for all the updates!

> a wild mix between "no argument", "utf-8", "UTF-8", and "utf-8".
`utf-8` *and* `utf-8`? :sweat_smile:

**Comment by martinpitt** [comment]
Ah, I suppose I wanted to write "utf8", and then saw that we don't actually use this form (it works, too, though). :grin: Anyway, harmless enough..
